src/relay/qnn/op/subtract.cc

```diff
@@ -0,0 +1,103 @@
+                             const Array<Expr> &new_args,
+                             const Array<tvm::relay::Type> &arg_types) {
+  // Get the args.
+  QnnBinaryOpArguments args(new_args);
+
+  // Get the input dtype and shape.
+  QnnBinaryOpTensorType input_type(arg_types, 0);
+
+  // TODO(shoubhik) - The lowering can be further optimized. Instead of inserting requantize in
+  // the start, we can insert requantize at the end if both input tensors have same qnn params. In
+  // that case, we can first subtract the tensors, add the zero point, and requantize at the end.
+  // This can be done in future.
+
+  // Since the input qnn params can be different than output qnn params, we first requantize the
+  // input tensors to the output qnn params. Then we call relay.subtract on the requantized inputs.
+  // This subtraction results in extra subtraction of the output zero point. We further add
+  // the zero point. The whole process can be represented using following equations
+  //
+  //          scale_c * (Q_c - zp_c) = scale_a * (Q_a - zp_a) - scale_b * (Q_b - zp_b)
+  //
+  // After requantizing Q_a and Q_b, equation becomes,
+  //          scale_c * (Q_c - zp_c) = scale_c * (Q_a' - zp_c) - scale_c * (Q_b' - zp_c)
+  //          scale_c * (Q_c - zp_c) = scale_c * (Q_a' - Q_b')
+  //
+  // Comparing the LHS and RHS, it results in
+  //          Q_c = Q_a' - Q_b' + zp_c
+  // The subtract op is done in int32 precision.
+
+  // Requantize LHS if necessary. Computes Q_a'
+  auto requantized_lhs = RequantizeOrUpcast(args.lhs, args.lhs_scale,
+                                            args.lhs_zero_point,
+                                            args.output_scale,
+                                            args.output_zero_point,
+                                            input_type.shape);
+  // Requantize RHS if necessary. Computes Q_b'
+  auto requantized_rhs = RequantizeOrUpcast(args.rhs, args.rhs_scale,
+                                            args.rhs_zero_point,
+                                            args.output_scale,
+                                            args.output_zero_point,
+                                            input_type.shape);
+
+  // Computes Q_a' - Q_b'
+  auto output = Subtract(requantized_lhs, requantized_rhs);
+
+  // Add zero point. Computes (Q_a' - Q_b') + zp_c
+  auto zero_scalar = MakeConstantScalar(DataType::Int(32), 0);
+  if (!IsEqualScalar(args.output_zero_point, zero_scalar)) {
+    output = Add(output, args.output_zero_point);
+  }
+
+  // Go back to lower precision.
+  return ConvertDtype(output, input_type.dtype);
+}
+
+// QNN Addition operator.
+QNN_REGISTER_BINARY_OP("subtract")
+.describe("Elementwise subtract with with broadcasting for quantized tensors.")
+.set_support_level(11)
+.set_attr<FTVMLegalize>("FTVMQnnCanonicalize", QnnSubtractCanonicalize)
+.set_attr<FInferCorrectLayout>("FInferCorrectLayout", QnnBinaryBroadcastLayout);
```

Comment:
While you are here, this can be moved to op_common.h where we define MACRO `QNN_REGISTER_BINARY_OP`, and remove it from here.

Reply:
I beleive this should be here. otherwise there woluld be a confusion as to where the op is registered. I have seen this convention being followed in all other cc files. I think we should stick with the convention. plus, there is a danger of moving this to .h file. just is case someone screws up later and removes or forgets to link the header file these ops will never be registerd.

Reply:
I meant here 

https://github.com/apache/incubator-tvm/blob/949dca4d078dbeb7990294b1f7a83de5b6ed4a25/src/relay/qnn/op/op_common.h#L86


Just like we share InferType function, we can share InferLayout between add/mul/subtract.

Reply:
The suggested change is only the line 99 

~~~
.set_attr<FInferCorrectLayout>("FInferCorrectLayout", QnnBinaryBroadcastLayout);

~~~

Reply:
done